prereise/gather/griddata/hifld/data_process/topology.py

```diff
@@ -0,0 +1,119 @@
+from itertools import combinations, product
+
+import networkx as nx
+from powersimdata.utility.distance import haversine
+from tqdm import tqdm
+
+from prereise.gather.griddata.hifld.const import abv_state_neighbor
+
+
+def min_dist_of_2_conn_comp(nodes1, nodes2, dist_metric):
+    """Calculate the minimum distance of two connected components based on the given
+    distance metric.
+
+    :param pandas.DataFrame nodes1: a data frame contains all the nodes of the first
+        connected component with index being node ID and two columns being latitude
+        and longitude in order.
+    :param pandas.DataFrame nodes2: a data frame contains all the nodes of the second
+        connected component with index being node ID and two columns being latitude
+        and longitude in order.
+    :param func dist_metric: a function defines the distance metric.
+    :return: (*tuple*) -- a tuple contains three elements, the minimum distance
+        follows by a pair of nodes from the two connected components respectively
+        in order
+    """
+    min_dist = float("inf")
+    for v1, v2 in product(nodes1.index, nodes2.index):
+        tmp_dist = dist_metric(nodes1.loc[v1], nodes2.loc[v2])
+        if tmp_dist < min_dist:
+            min_dist, res_v1, res_v2 = tmp_dist, v1, v2
```

Comment:
Can we potentially speed this up with chained `apply` calls to fill a dataframe and then the use of `min` & `idxmin` to select the winners?

Reply:
I did following test and they looks similar:

![image](https://user-images.githubusercontent.com/52716585/132065524-aa6b08f2-f71a-4495-a519-38f675c743ed.png)

Potential reasons of `apply` being slower is calculating `min` and `idxmin` (not included in the test above yet) needs another two sweep of O(N).

Reply:
I did following test and they looks similar:
![image](https://user-images.githubusercontent.com/52716585/132065741-c49f1cfd-de52-4544-926d-f09d0c98c26a.png)
Potential reasons of apply being slower is calculating ``min`` and
idxmin
(not included in the test above yet) needs another two sweep of O(N).

Reply:
I think we only want to call `min` and `idxmin` on a full dataframe that we construct with the chained applies, not within one of the applies to end up with a Series. We may still not achieve any/much speedup though.

Reply:
Running the following code to simulate the heaviest computation combination, I see a ~6x speedup with an `apply` version:
```
cc1 = all_cc[-1]
cc2 = all_cc[-2]
nodes1 = subs.loc[subs.STATE.isin({"ME", "NH"}) & subs.index.isin(cc1), ["LATITUDE", "LONGITUDE"]]
nodes2 = subs.loc[cc2, ["LATITUDE", "LONGITUDE"]]
start_time = timeit.default_timer()
min_dist, sub1, sub2 = min_dist_of_2_conn_comp(nodes1, nodes2, haversine)
print(timeit.default_timer() - start_time)
start_time = timeit.default_timer()
distances = nodes1.apply(lambda x: nodes2.apply(lambda y: haversine(x, y), axis=1), axis=1)
nodes1_min = distances.min(axis=1).idxmin()
nodes2_min = distances.loc[nodes1_min].idxmin()
distance_min = distances.loc[nodes1_min, nodes2_min]
print(timeit.default_timer() - start_time)
```
The first one executes in 1.2-1.3 seconds on my machine, compared to 0.2 seconds for the second one.

Reply:
If we don't calculate minimum on the fly, that's indeed faster. This is actually my previous implementation. I wasn't able to test the computational performance of it due to running out of memory (with other stuffs running in parallel). I think here is a clear trade-off between time complexity and memory complexity. I'm fine with either one. What do you think @rouille @jon-hagg ?

Reply:
If we don't filter the big CC to just the appropriate states, we see a similar level of speedup:
```python
>>> cc1 = all_cc[-1]
>>> cc2 = all_cc[-2]
>>> nodes1 = subs.loc[cc1, ["LATITUDE", "LONGITUDE"]]
>>> nodes2 = subs.loc[cc2, ["LATITUDE", "LONGITUDE"]]
>>> start_time = timeit.default_timer()
>>> min_dist, sub1, sub2 = min_dist_of_2_conn_comp(nodes1, nodes2, haversine)
>>> print(timeit.default_timer() - start_time)
185.3390242999999
>>> start_time = timeit.default_timer()
>>> distances = nodes1.apply(lambda x: nodes2.apply(lambda y: haversine(x, y), axis=1), axis=1)
>>> nodes1_min = distances.min(axis=1).idxmin()
>>> nodes2_min = distances.loc[nodes1_min].idxmin()
>>> distance_min = distances.loc[nodes1_min, nodes2_min]
>>> print(timeit.default_timer() - start_time)
27.511743399999887
```

Even in this intentionally inefficient case, the `distances` data frame is dimension (50607, 41), which for a float64 dtype seems to be roughly 17 MB (per some back-of-the-envelope calculations).

Reply:
We have going to have hundreds of such data frame though unless we clean up memory on the fly.

Reply:
I'm not an expert in memory management, but won't the old data frame be automatically discarded (and its memory freed) after the function which creates it returns, or the next time that we re-define a new object into that variable name? I think python's garbage collector should handle that automatically.

Reply:
I'm not an expert either! Python indeed automatically release memory after function call but not all of them, even in the case an object loses references completely. The logic behind that could be complex, my quick googling finds following [post](https://stackoverflow.com/questions/15455048/releasing-memory-in-python). 

The story is I ran out of memory in a resource limited situation during implementation, thus, I switched to another one. I have no strong preferences in such trade-offs.

Reply:
I've added an optional parameter `memory_efficient` for function `min_dist_of_2_conn_comp` and the user facing function `connect_islands_with_minimum_cost` which passes the parameter through. Once the parameter is set, the current implementation is triggered. Otherwise, the default behavior is using `apply`.